doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,270 @@
+# Materialize Transactional Consistency
+
+## Summary
+
+A description of Materialize’s consistency guarantees with respect to the operations that occur inside of Materialize.
+Operations include: `SELECT`, `INSERT`, `UPDATE`, and `DELETE` statements (but not `TAIL`)
+
+This design document is in service of [Epic #11631](https://github.com/MaterializeInc/materialize/issues/11631)
+and [Correctness Property #1](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/platform/ux.md#correctness)
+(copied below).
+
+> *Transactions in Materialize are strictly serializable with respect to the operations that occur inside of Materialize.* Operations include:
+> - `SELECT`, `INSERT`, `UPDATE`, and `DELETE` statements (but not `TAIL`)
+> - Materialize initiated acknowledgements of upstream sources (*e.g., the commit of an offset by a Kafka source, the acknowledge of an LSN by a PostgresSQL source, etcetera*)
+
+This document does not take any opinion on whether the changes proposed should be the default behavior or an opt-in
+behavior via some configuration.
+
+Time units of milliseconds are used below for convenience to reflect current implementations. However, this document
+does not take any opinion on what time units should be used for the design proposed.
+
+## Goals
+
+The goal of this design document is to describe what changes are needed to provide Strict Serializability to the
+operations mentioned in correctness property 1.
+
+## Non-Goals
+
+The following items' impact on consistency is not considered:
+
+- `TAIL`
+- Sinks.
```

Comment:
I do wonder if the _initial_ timestamp chosen for `TAIL`/sinks should obey strict serializability. Like, if you start a `TAIL`, seems like it'd be nice to guarantee that the initial snapshot will contain all writes you've committed.